Add documents concurrency fix

pandu-k opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
During the add documents call, if another user deletes the index, the add_documents user gets a 500 due to a keyerror during the mappings update.

* **What is the new behavior (if this is a feature change)?**
Now, the add_documents call attempts to find index mappings from Marqo-os if it isn't in the cache. If the index has been deleted in this time, the add_documents user will get "IndexNotFound" error. 

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
In progress

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

pandu-k committed:
Created fix and tests

wanliAlex commented on src/marqo/tensor_search/backend.py:
```diff
@@ -108,7 +108,7 @@ def add_customer_field_properties(config: Config, index_name: str,
         }
     }
-    existing_info = get_cache()[index_name]
+    existing_info = get_index_info(config=config, index_name=index_name)
```
to what extent will this change slow the `add_documents` process?

pandu-k replied:
thanks - updated.

wanliAlex commented on tests/tensor_search/test_index_meta_cache.py:
```diff
@@ -577,3 +579,78 @@ def threaded_while(thread_num, loop_record):
+                "docs": [
+                    {"Title": "Blah"}, {"Title": "blah2"}, {"Title": "Blah3"}, {"Title": "Blah4"},
+                ]
+            })
```
do we need a `clear_cache_thread.join()` to make it safer?

pandu-k replied:
As in to wait for the thread to complete before continuing on with the next tests? I don't know how this will make it safer

wanliAlex commented on tests/tensor_search/test_index_meta_cache.py:
```diff
@@ -577,3 +579,78 @@ def threaded_while(thread_num, loop_record):
+            return True
+        assert run()
+
+    def test_add_documents_to_non_existent_index(self):
```
in general, under what circumstance will an `index_cached` get deleted but the `index` still exists?

pandu-k replied:
This should be quite rare:
1. thread_a does add_documents
2. thread_a checks the cache. the index is there
3. thread_b deletes the index. the index is removed from cache
4. thread_c, on another Marqo instance, creates an index with the same name
5. thread_a checks cache at the end of add_documents. The index is not there, but does actually exist

This is simulated in `test_add_documents_to_unknown_index()`

pandu-k committed:
existing_info set to get cached index info

made a longer sleep in index_meta_cache (as search refreshes index info every 2 secs)
